client/core/bond.go

```diff
@@ -39,58 +40,217 @@ func (c *Core) watchExpiredBonds(ctx context.Context) {
+// this estimate, but it is still very short relative to the entire bond
+// lifetime, which is on the order of months.
+func spendableDelay(net dex.Network) int64 {
+	// We use 3*pendingBuffer as a well-padded estimate of this duration. e.g.
+	// with mainnet, we would use a 90 minute pendingBuffer and a 270 minute
+	// spendableDelay to be robust to periods of slow blocks.
+	return 3 * pendingBuffer(net)
+}
+
+// minBondLifetime gives the minimum bond lifetime (a duration from now until
+// lockTime) that a new bond should use to prevent a tier drop. bondExpiry is in
+// seconds.
+func minBondLifetime(net dex.Network, bondExpiry int64) time.Duration {
+	lifeTimeSec := 2*(pendingBuffer(net)+bondExpiry) + spendableDelay(net) // 2*(p+E)+m
+	return time.Second * time.Duration(lifeTimeSec)
+}
+
+func (c *Core) rotateBonds(ctx context.Context) {
 	// 1. Refund bonds with passed lockTime.
 	// 2. Move bonds that are expired according to DEX bond expiry into
 	//    expiredBonds (lockTime<lockTimeThresh).
-	// 3. TODO: Add bonds to keep N bonds active, according to some optional config.
+	// 3. Add bonds to keep N bonds active, according to target tier and max
+	//    bonded amount, posting before expiry of the bond being replaced.
 	now := time.Now().Unix()
 
 	type bondID struct {
 		assetID uint32
 		coinID  []byte
 	}
 
+	bondKey := c.bondKey()
+	if bondKey != nil {
+		defer bondKey.Zero() // it's a copy
+	}
+
 	for _, dc := range c.dexConnections() {
+		// time.Now().After(time.Unix(bond.LockTime-dc.cfg.BondExpiry))
 		lockTimeThresh := now // in case dex is down, expire (to refund) when lock time is passed
+		var bondExpiry int64
+		bondAmts := make(map[uint32]uint64)
+		bondAssets := make(map[uint32]*msgjson.BondAsset)
 		dc.cfgMtx.RLock()
 		if dc.cfg != nil {
+			bondExpiry = int64(dc.cfg.BondExpiry)
+			for symb, ba := range dc.cfg.BondAssets {
+				id, _ := dex.BipSymbolID(symb)
+				bondAmts[id] = ba.Amt
+				bondAssets[id] = ba
+			}
 			lockTimeThresh += int64(dc.cfg.BondExpiry)
 		}
 		dc.cfgMtx.RUnlock()
+		replaceThresh := lockTimeThresh + pendingBuffer(c.net) // replace before expiry to avoid tier drop
 
-		filterExpiredBonds := func(bond []*db.Bond) (liveBonds []*db.Bond) {
-			for _, bond := range bond {
+		var weak []*db.Bond // may re-post ahead of expiry to maintain tier
+		filterExpiredBonds := func(bonds []*db.Bond) (liveBonds []*db.Bond) {
+			for _, bond := range bonds {
 				if int64(bond.LockTime) <= lockTimeThresh {
 					// This is generally unexpected because auth, reconnect, or a
 					// bondexpired notification should do this first.
 					dc.acct.expiredBonds = append(dc.acct.expiredBonds, bond)
-					c.log.Warnf("Expired bond found: %v", coinIDString(bond.AssetID, bond.CoinID))
+					c.log.Infof("Newly expired bond found: %v (%s)", coinIDString(bond.AssetID, bond.CoinID), unbip(bond.AssetID))
 				} else {
+					if int64(bond.LockTime) <= replaceThresh {
+						weak = append(weak, bond) // but still live
+					}
```

Comment:
The bond may be pending. Does that require special handling?

Reply:
Good question.  For purposes of `mustPost` computation, I believe this comes out correctly:

```go
		tierDeficit := int64(targetTier) - tier
		mustPost := tierDeficit + weakStrength - pendingStrength
```

By putting that pending bond in the weak bucket also (note it falls through and stays in `pendingBonds` too), we negate the pending bond's offset to `mustPost`, which I believe is correct.

Will consider that more however.